lib/routes/account.js

```diff
@@ -406,52 +412,55 @@ module.exports = function (
-                  tokenVerificationId = undefined
-                  mustVerifySession = false
-                  doSigninConfirmation = false
-                } else {
-                  // The user doesn't *have* to verify their session if they're not requesting keys,
-                  // but we still create it with a non-null tokenVerificationId, so it will still
-                  // be considered unverified.  This prevents the session from being used for sync
-                  // unless the user explicitly requests us to resend the confirmation email, and completes it.
-                  mustVerifySession = requestHelper.wantsKeys(request)
-                  doSigninConfirmation = mustVerifySession && emailRecord.emailVerified
-                }
-
-                if(email !== emailRecord.email) {
-                  customs.flag(request.app.clientAddress, {
-                    email: email,
-                    errno: error.ERRNO.INCORRECT_PASSWORD
-                  })
-                  throw error.incorrectPassword(emailRecord.email, email)
-                }
-
-                return checkPassword(emailRecord, authPW, request.app.clientAddress)
-                  .then(
-                    function (match) {
-                      if (! match) {
-                        throw error.incorrectPassword(emailRecord.email, email)
+                allowSigninUnblock = requestHelper.shouldEnableSigninUnblock(emailRecord, config, request)
+                if (allowSigninUnblock && unblockCode) {
+                  return db.consumeUnblockCode(emailRecord.uid, unblockCode)
+                    .then(
+                      function (code) {
+                        if (Date.now() - code.createdAt > unblockCodeLifetime) {
+                          log.info({
+                            op: 'Account.login.unblockCode.expired',
+                            uid: emailRecord.uid.toString('hex')
+                          })
+                          throw error.invalidUnblockCode()
+                        }
+                        didSigninUnblock = true
                       }
-
-                      return log.activityEvent('account.login', request, {
-                        uid: emailRecord.uid.toString('hex')
-                      })
-                    }
-                  )
+                    )
+                    .catch(
+                      function (err) {
+                        if (err.errno === error.ERRNO.UNBLOCK_CODE_INVALID) {
+                          customs.flag(request.app.clientAddress, {
+                            email: email,
+                            errno: err.errno
+                          })
+                        }
+                        throw err
+                      }
+                    )
+                } else if (!allowSigninUnblock && maybeBlocked) {
+                  throw maybeBlocked
+                }
```

Comment:
I wonder if it's worth explicitly erroring out if `allowSigninUnblock=false` but there's an `unblockCode` in the request; IIUC this should never happen in practice, unless perhaps you're signing in right in the middle of a deploy that changs whether or not you're in the `allowSigninUnblock` group.

Also, it took me a little while to process the flow of control between the two branches of this `if` statement, and the `checkUnblockCode` function.  I wonder if rather than an if-then-else, it would be clearer to have something like:

```
if (allowSigninUnblock && unblockCode) {
  ... existing logic ...
}
if (maybeBlocked && !didSigninUnblock) {
  throw maybeBlocked
}
```

This would make it more obvious that there aren't any unhandled paths through the branching conditions, and that `maybeBlocked` is always re-thrown if you didn't complete the unblocking.

Just a suggestion based on working through the flow, I don't feel super strongly about it either way.